test/performance/search/dream_index/binning_directory_benchmark.cpp

```diff
@@ -0,0 +1,144 @@
+                              benchmark::Counter::OneK::kIs1000);
+}
+
+static void arguments(benchmark::internal::Benchmark* b)
+{
+    for (int32_t bins : {64, 8192})
+    {
+        for (int32_t bits = 1<<15; bits <= 1<<25; bits <<= 5)
+        {
+            for (int32_t hash_num = 2; hash_num < 4; ++hash_num)
+            {
+                b->Args({bins, bits, hash_num, 100'000});
+            }
+        }
+    }
+}
+
+inline std::vector<size_t> generate_numbers(size_t const len = 500,
+                                            size_t const min = std::numeric_limits<size_t>::lowest(),
+                                            size_t const max = std::numeric_limits<size_t>::max(),
+                                            size_t const seed = 0)
+{
+    std::mt19937 engine(seed);
+    std::uniform_int_distribution<size_t> dist{min, max};
+
+    auto gen = [&dist, &engine]() { return dist(engine); };
+    std::vector<size_t> sequence(len);
+    std::generate(std::ranges::begin(sequence), std::ranges::end(sequence), gen);
+
+    return sequence;
+}
+
+
+template <typename t>
+class bd_benchmark : public benchmark::Fixture
+{
+public:
+    size_t bins;
+    size_t bits;
+    size_t num_hash;
+    size_t sequence_length;
+    std::vector<size_t> hash_values;
+    std::vector<size_t> bin_values;
+    t bd;
+    sdsl::bit_vector res;
+
+    void SetUp(::benchmark::State const & state)
+    {
+        bins = static_cast<size_t>(state.range(0));
+        bits = static_cast<size_t>(state.range(1));
+        num_hash = static_cast<size_t>(state.range(2));
+        sequence_length = static_cast<size_t>(state.range(3));
+
+        bin_values  = generate_numbers(sequence_length, 0u, bins - 1);
+        hash_values = generate_numbers(sequence_length);
+        binning_directory tmp(bins, bits, num_hash);
+        bd = t(std::move(tmp));
+        res.resize(bins);
+    }
+
+    void TearDown(::benchmark::State const &) {}
+
+    void set_benchmark(::benchmark::State & state)
+    {
+        for (auto _ : state)
+        {
+            for (auto [hash, bin] : std::view::zip(hash_values, bin_values))
+                bd.set(hash, bin);
+        }
+
+        state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
+    }
+
+    void get_benchmark(::benchmark::State & state)
+    {
+        for (auto _ : state)
+        {
+            for (auto hash : hash_values)
+                res = bd.get(hash, std::move(res));
+        }
+
+        state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
+    }
+
+    void get_copy_benchmark(::benchmark::State & state)
+    {
+        for (auto _ : state)
+        {
+            for (auto hash : hash_values)
+                res = bd.get(hash);
+        }
+
+        state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
+    }
+};
+
+BENCHMARK_TEMPLATE_DEFINE_F(bd_benchmark, set, binning_directory) (benchmark::State & state)
```

Comment:
This looks over complicated..

I think it can be replaced with:

```cpp

auto set_up(size_t bins, size_t bits, size_t num_hash, size_t sequence_length)
{
    bin_values  = generate_numbers(sequence_length, 0u, bins - 1);
    hash_values = generate_numbers(sequence_length);
    binning_directory tmp(bins, bits, num_hash);
    bd = t(std::move(tmp));
    res.resize(bins);
    
    return {bin_values, hash_values, bd};
}

template <typename binning_directory_t>
void set_benchmark(::benchmark::State & state)
{
    auto && [ bin_values, hash_values, bd ] = set_up(state.range(0), state.range(1), state.range(2), state.range(3));
    
    for (auto _ : state)
    {
        for (auto [hash, bin] : std::view::zip(hash_values, bin_values))
            bd.set(hash, bin);
    }

    state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
}

template <typename binning_directory_t>
void get_benchmark(::benchmark::State & state)
{
    auto && [ bin_values, hash_values, bd ] = set_up(state.range(0), state.range(1), state.range(2), state.range(3));
    
    for (auto _ : state)
    {
        for (auto hash : hash_values)
            res = bd.get(hash, std::move(res));
    }

    state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
}

template <typename binning_directory_t>
void get_copy_benchmark(::benchmark::State & state)
{
    auto && [ bin_values, hash_values, bd ] = set_up(state.range(0), state.range(1), state.range(2), state.range(3));
    
    for (auto _ : state)
    {
        for (auto hash : hash_values)
            res = bd.get(hash);
    }

    state.counters["hashes/sec"] = hashes_per_second(std::ranges::size(hash_values));
}

BENCHMARK_TEMPLATE(set_benchmark, binning_directory)->Apply(arguments);
BENCHMARK_TEMPLATE(set_benchmark, binning_directory_compressed)->Apply(arguments);

BENCHMARK_TEMPLATE(get_benchmark, binning_directory)->Apply(arguments);
BENCHMARK_TEMPLATE(get_benchmark, binning_directory_compressed)->Apply(arguments);

BENCHMARK_TEMPLATE(get_copy_benchmark, binning_directory)->Apply(arguments);
BENCHMARK_TEMPLATE(get_copy_benchmark, binning_directory_compressed)->Apply(arguments);

```